src/java/picard/vcf/GenotypeConcordanceCounts.java

```diff
@@ -74,6 +85,76 @@ public void validateCountsAgainstScheme(final GenotypeConcordanceScheme scheme)
     }
 
     /**
+     * Genotype Concordance is the number of times the truth and call states match exactly / all truth and call combinations made
+     * If the GA4GH scheme is being used, any MISSING sites in truth OR call will not be included in the discordance calculations.
+     */
+    public double calculateGenotypeConcordance(final GenotypeConcordanceScheme scheme, final boolean missingSitesFlag) {
+        double numerator = 0.0;
+        double denominator = 0.0;
+
+        scheme.validateScheme();
+
+        final TruthState[] allTruthStates = TruthState.values();
+        final CallState[] allCallStates = CallState.values();
+
+        for (final TruthState truthState : allTruthStates) {
+            for (final CallState callState : allCallStates) {
+                if(!missingSitesFlag && isMissing(truthState, callState)) {
+                    continue;
+                }
+                else {
+                    final TruthAndCallStates truthAndCallStates = new TruthAndCallStates(truthState, callState);
+                    final int count = getCount(truthAndCallStates);
+                    if (truthState.getCode()==callState.getCode()) {
+                        //Because the TruthState and CallState enums are in the same order, the ordinals correspond to the same constants
+                        //If we enter this, we are 'on the diagonal
+                        numerator += count;
+                        denominator += count;
+                    } else {
+                        denominator += count;
+                    }
+                }
+            }
+        }
+        return (numerator / denominator);
```

Comment:
Even if there were only missing missing states, these would be in the numerator and the denominator so it should be 1. Unless it wasn't NIST then it would be 0/0.. So we could protect for that. 



> On Jul 31, 2015, at 12:31 AM, Eric Banks <notifications@github.com> wrote:
> 
> In src/java/picard/vcf/GenotypeConcordanceCounts.java:
> 
> > +                }
> > +                else {
> > +                    final TruthAndCallStates truthAndCallStates = new TruthAndCallStates(truthState, callState);
> > +                    final int count = getCount(truthAndCallStates);
> > +                    if (truthState.getCode()==callState.getCode()) {
> > +                        //Because the TruthState and CallState enums are in the same order, the ordinals correspond to the same constants
> > +                        //If we enter this, we are 'on the diagonal
> > +                        numerator += count;
> > +                        denominator += count;
> > +                    } else {
> > +                        denominator += count;
> > +                    }
> > +                }
> > +            }
> > +        }
> > +        return (numerator / denominator);
> Is it possible to get to this line with a denominator that has never been incremented (e.g. if there are only missing-missing states)? If so, the denominator will be zero and you'll exception out (division by zero is bad).
> You could protect yourself by doing this instead:
> return (denominator > 0.0 ? (numerator / denominator) : 0.0);
> 
> —
> Reply to this email directly or view it on GitHub.
>